File: src/test/java/cube8540/oauth/authentication/oauth/client/domain/OAuth2ClientTest.java
package cube8540.oauth.authentication.oauth.client.domain;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.oauth2.core.AuthorizationGrantType;

import static cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientTestsHelper.ADDED_SCOPE;
import static cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientTestsHelper.RAW_CHANGE_SECRET;
import static cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientTestsHelper.RAW_CLIENT_ID;
import static cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientTestsHelper.RAW_ENCODING_SECRET;
import static cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientTestsHelper.RAW_SECRET;
import static cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientTestsHelper.REDIRECT_URI;
import static cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientTestsHelper.makeErrorValidatorFactory;
import static cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientTestsHelper.makePassValidatorFactory;
import static cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientTestsHelper.makePasswordEncoder;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.when;

@DisplayName("OAuth2 클라이언트 테스트")
class OAuth2ClientTest {

    private OAuth2Client client;

    @BeforeEach
    void setup() {
        this.client = new OAuth2Client(RAW_CLIENT_ID, RAW_SECRET);
    }

    @Test
    @DisplayName("리다이렉트 URI 추가")
    void addRedirectUri() {
        this.client.addRedirectUri(REDIRECT_URI);

        assertTrue(client.getRedirectUris().contains(REDIRECT_URI));
    }

    @Test
    @DisplayName("리다이렉트 URI 삭제")
    void removeRedirectUri() {
        this.client.addRedirectUri(REDIRECT_URI);

        this.client.removeRedirectUri(REDIRECT_URI);

        assertFalse(client.getRedirectUris().contains(REDIRECT_URI));
    }

    @Test
    @DisplayName("클라이언트 인증 방식 추가")
    void addClientGrantType() {
        this.client.addGrantType(AuthorizationGrantType.AUTHORIZATION_CODE);

        assertTrue(client.getGrantTypes().contains(AuthorizationGrantType.AUTHORIZATION_CODE));
    }

    @Test
    @DisplayName("클라이언트 인증 방식 삭제")
    void removeClientGrantType() {
        this.client.addGrantType(AuthorizationGrantType.AUTHORIZATION_CODE);

        this.client.removeGrantType(AuthorizationGrantType.AUTHORIZATION_CODE);
        assertFalse(client.getGrantTypes().contains(AuthorizationGrantType.AUTHORIZATION_CODE));
    }

    @Test
    @DisplayName("클라이언트 스코프 추가")
    void addClientScope() {
        this.client.addScope(ADDED_SCOPE);

        assertTrue(client.getScopes().contains(ADDED_SCOPE));
    }

    @Test
    @DisplayName("클라이언트 스코프 삭제")
    void removeClientScope() {
        this.client.addScope(ADDED_SCOPE);

        this.client.removeScope(ADDED_SCOPE);
        assertFalse(client.getScopes().contains(ADDED_SCOPE));
    }

    @Test
    @DisplayName("클라이언트 패스워드 암호화")
    void clientSecretEncrypting() {
        PasswordEncoder encoder = makePasswordEncoder(RAW_SECRET, RAW_ENCODING_SECRET);

        this.client.encrypted(encoder);
        assertEquals(RAW_ENCODING_SECRET, client.getSecret());
    }

    @Test
    @DisplayName("유효 하지 않은 정보가 저장 되어 있을시")
    void clientDataInvalid() {
        OAuth2ClientValidatorFactory factory = makeErrorValidatorFactory(client);

        assertThrows(ClientInvalidException.class, () -> client.validate(factory));
    }

    @Test
    @DisplayName("모든 데이터가 유효할시")
    void clientDataAllowed() {
        OAuth2ClientValidatorFactory factory = makePassValidatorFactory(client);

        assertDoesNotThrow(() -> client.validate(factory));
    }

    @Test
    @DisplayName("이전에 사용 하던 패스워드와 서로 일치 하지 않을시")
    void whenNotMatchedExistsSecret() {
        PasswordEncoder encoder = makePasswordEncoder(RAW_SECRET, RAW_ENCODING_SECRET);

        this.client.encrypted(encoder);
        when(encoder.matches(RAW_SECRET, RAW_ENCODING_SECRET)).thenReturn(false);

        ClientAuthorizationException e = assertThrows(ClientAuthorizationException.class, () -> client.changeSecret(RAW_SECRET, RAW_CHANGE_SECRET, encoder));
        assertEquals(ClientErrorCodes.INVALID_PASSWORD, e.getCode());
    }

    @Test
    @DisplayName("패스워드 변경")
    void changeSecret() {
        PasswordEncoder encoder = makePasswordEncoder(RAW_SECRET, RAW_ENCODING_SECRET);

        this.client.encrypted(encoder);

        client.changeSecret(RAW_SECRET, RAW_CHANGE_SECRET, encoder);
        assertEquals(RAW_CHANGE_SECRET, client.getSecret());
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/client/domain/OAuth2ClientTest.kt
package cube8540.oauth.authentication.oauth.client.domain

import cube8540.validator.core.ValidationError
import cube8540.validator.core.ValidationRule
import cube8540.validator.core.Validator
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.*
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.security.crypto.password.PasswordEncoder

class OAuth2ClientTest {

    private val clientId = "clientId"
    private val clientSecret = "clientSecret"

    @Nested
    inner class EncryptingTest {

        private val client = OAuth2Client(clientId, clientSecret)

        @Test
        fun `encoding client secret`() {
            val encoder: PasswordEncoder = mockk()

            every { encoder.encode("clientSecret") } returns "encodedClientSecret"

            client.encrypted(encoder)
            assertThat(client.secret).isEqualTo("encodedClientSecret")
        }
    }

    @Nested
    inner class ValidationTest {

        private val validatorFactory: OAuth2ClientValidatorFactory = mockk()
        private val client = OAuth2Client(clientId, clientSecret)

        @Test
        fun `client data is invalid`() {
            val rule: ValidationRule<OAuth2Client> = mockk {
                every { isValid(client) } returns false
                every { error() } returns ValidationError("clientId", "test")
            }
            every { validatorFactory.createValidator(client) } returns Validator.of(client).registerRule(rule)

            val thrown = catchThrowable { client.validate(validatorFactory) }
            assertThat(thrown).isInstanceOf(ClientInvalidException::class.java)
        }

        @Test
        fun `client data is allowed`() {
            val rule: ValidationRule<OAuth2Client> = mockk {
                every { isValid(client) } returns true
            }
            every { validatorFactory.createValidator(client) } returns Validator.of(client).registerRule(rule)

            assertThatCode { client.validate(validatorFactory) }.doesNotThrowAnyException()
        }
    }

    @Nested
    inner class SecretChangeTest {

        private val client = OAuth2Client(clientId, clientSecret)

        @Test
        fun `existing secret is not equal`() {
            val encoder: PasswordEncoder = mockk {
                every { matches("existingSecret", clientSecret) } returns false
            }

            val thrown = catchThrowable { client.changeSecret("existingSecret", "newSecret", encoder) }
            assertThat(thrown).isInstanceOf(ClientAuthorizationException::class.java)
            assertThat((thrown as ClientAuthorizationException).code).isEqualTo(ClientErrorCodes.INVALID_PASSWORD)
        }

        @Test
        fun `change secret successful`() {
            val encoder: PasswordEncoder = mockk {
                every { matches("existingSecret", clientSecret) } returns true
            }

            client.changeSecret("existingSecret", "newSecret", encoder)
            assertThat(client.secret).isEqualTo("newSecret")
        }
    }
}